### PR TITLE
added support to generate primitives from log

### LINKED ISSUE #110: generate the primitives as per the JSON file in store
## Feature Description
- add functionality like `provide_primitives(log_file_ctypto_store)`  which generates the primitives and stores them in the CryptoStore.
- `provide_primitives(log_file or dictionary with primitive_name, shape, nr_instances)`

### PR DESCRIPTION
**Pull request by aanurraj**
## Description
this PR closes #110 
made few changes in generating logs will be discussed further with @gmuraru 

## Affected Dependencies
None

## How has this been tested?
Tested on local, tests are still pending.

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [x] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [x] My changes are covered by tests

### CODE REVIEW

File: src/sympc/store/crypto_primitive_provider.py [hunk 1]
```diff
@@ -110,11 +111,40 @@ def stop_logging(generate_file: bool = False):
             json: returns the json object containing ops details.
         """
         CryptoPrimitiveProvider._LOGGING = False
-        log_json = json.dumps(CryptoPrimitiveProvider._ops_list)
+        log = copy(CryptoPrimitiveProvider._ops_list)
```
Comment (gmuraru) [inline src/sympc/store/crypto_primitive_provider.py:114]: MAYBE: Use ```CryptoPrimitiveProvider._ops_list.copy()``` (in that case we don't need the import)

File: tests/sympc/store/crypto_primitive_provider_test.py [hunk 2]
```diff
@@ -191,10 +186,16 @@ def test_primitive_logging_model(get_clients) -> None:
...
     "ops",
     [
-        ["beaver_mul", {"a_shape": [1, 5], "b_shape": [1, 5]}],
-        ["beaver_matmul", {"a_shape": [1, 2880], "b_shape": [2880, 10]}],
-        ["beaver_conv2d", {"a_shape": [1, 1, 28, 28], "b_shape": [5, 1, 5, 5]}],
-        ["fss_comp", {}],
+        ["beaver_mul", {"a_shape": [1, 5], "b_shape": [1, 5], "nr_parties": 2}],
+        [
+            "beaver_matmul",
+            {"a_shape": [1, 2880], "b_shape": [2880, 10], "nr_parties": 2},
```
Comment (gmuraru) [inline tests/sympc/store/crypto_primitive_provider_test.py:192]: Q: Are the number of parties required?

Comment (aanurraj) [reply tests/sympc/store/crypto_primitive_provider_test.py:192]: yes, else we get this error:

`TypeError: _get_triples() missing 1 required positional argument: 'nr_parties'`

File: src/sympc/store/crypto_primitive_provider.py (hunk 1 above)
Comment (aanurraj) [reply src/sympc/store/crypto_primitive_provider.py:114]: working

File: tests/sympc/store/crypto_primitive_provider_test.py (hunk 2 above)
Comment (gmuraru) [reply tests/sympc/store/crypto_primitive_provider_test.py:192]: Ahh..it is ok how it is now :smile:

Comment (aanurraj) [reply tests/sympc/store/crypto_primitive_provider_test.py:192]: no problem :D

File: src/sympc/store/crypto_primitive_provider.py [hunk 3]
```diff
@@ -110,11 +110,40 @@ def stop_logging(generate_file: bool = False):
...
 
         if generate_file:
             with open(CryptoPrimitiveProvider._FILENAME, "w") as f:
-                f.write(log_json)
+                f.write(json.dumps(log))
+        return dict(log)
 
-        return log_json
+    @staticmethod
+    def generate_primitive_from_dict(primitive_log: dict, session: Session) -> None:
```
Comment (gmuraru) [inline src/sympc/store/crypto_primitive_provider.py:122]: NITPICK: Type hint for the dict.
```dict[str, Any```

File: src/sympc/store/crypto_primitive_provider.py (hunk 1 above)
Comment (aanurraj) [reply src/sympc/store/crypto_primitive_provider.py]: done

File: src/sympc/store/crypto_primitive_provider.py (hunk 3 above)
Comment (aanurraj) [reply src/sympc/store/crypto_primitive_provider.py]: donee

File: src/sympc/store/crypto_primitive_provider.py [hunk 4]
```diff
@@ -110,11 +110,35 @@ def stop_logging(generate_file: bool = False):
...
+            primitive_log (dict): the dict log created with primitive logging.
+            session (Session): Session.
+
+        Raises:
+            ValueError: If primitive_log is None.
+        """
+        if primitive_log is None:
+            raise ValueError("The provided log is cannot be None")
+
+        for op_str, args in primitive_log.items():
```
Comment (gmuraru) [inline src/sympc/store/crypto_primitive_provider.py:137]: NITPICK: You can do
```
for op_str, (g_kwargs, p_kwargs) in primitive_log.items():
```

File: tests/sympc/store/crypto_primitive_provider_test.py [hunk 5]
```diff
@@ -174,57 +170,203 @@ def test_primitive_logging_model(get_clients) -> None:
...
+
+    a_shape = (2, 3)
+    b_shape = (3, 10)
+
+    key = f"beaver_matmul_{a_shape}_{b_shape}"
+
+    store_client_1 = session.session_ptrs[0].crypto_store.store.get()
+    store_client_2 = session.session_ptrs[1].crypto_store.store.get()
+
+    shape_client_1 = tuple(store_client_1.get(key)[0][0].shape)
```
Comment (gmuraru) [inline tests/sympc/store/crypto_primitive_provider_test.py:285]: MAYBE: You also need to check the other shape, right?

Comment (aanurraj) [reply tests/sympc/store/crypto_primitive_provider_test.py:285]: other shape ? I did not get 😟

File: src/sympc/store/crypto_primitive_provider.py (hunk 4 above)
Comment (aanurraj) [reply src/sympc/store/crypto_primitive_provider.py:137]:
```
for op_str, args in primitive_log.items():
            for (p_kwargs, g_kwargs) in args:
```

value in `primitive_log.items` is a list

Comment (aanurraj) [reply src/sympc/store/crypto_primitive_provider.py:137]: thanks :D

File: tests/sympc/store/crypto_primitive_provider_test.py (hunk 5 above)
Comment (aanurraj) [reply tests/sympc/store/crypto_primitive_provider_test.py]: done

**Review by gmuraru — APPROVED** [review]
LGTM!

**Pull request merged**